packages/components/src/utils/hooks/extract-color.js

```diff
@@ -0,0 +1,211 @@
+  It also simplifies some of the canvas context manipulation
+  with a set of helper functions.
+*/
+
+const CanvasImage = function ( image ) {
+	this.canvas = document.createElement( 'canvas' );
+	this.context = this.canvas.getContext( '2d' );
+	this.width = this.canvas.width = image.width;
+	this.height = this.canvas.height = image.height;
+	this.context.drawImage( image, 0, 0, this.width, this.height );
+};
+
+CanvasImage.prototype.getImageData = function () {
+	return this.context.getImageData( 0, 0, this.width, this.height );
+};
+
+export const ColorThief = function () {};
+
+/*
+ * getColor(sourceImage[, quality])
+ * returns {r: num, g: num, b: num}
+ *
+ * Use the median cut algorithm provided by quantize.js to cluster similar
+ * colors and return the base color from the largest cluster.
+ *
+ * Quality is an optional argument. It needs to be an integer. 1 is the highest quality settings.
+ * 10 is the default. There is a trade-off between quality and speed. The bigger the number, the
+ * faster a color will be returned but the greater the likelihood that it will not be the visually
+ * most dominant color.
+ *
+ * */
+ColorThief.prototype.getColor = function ( sourceImage, quality = 10 ) {
+	const palette = this.getPalette( sourceImage, 5, quality );
+	/**
+	 * Custom update:
+	 * The palette may be null if the image is pure white.
+	 */
+	const dominantColor = palette ? palette[ 0 ] : undefined;
+	return dominantColor;
+};
+
+/*
+ * getPalette(sourceImage[, colorCount, quality])
+ * returns array[ {r: num, g: num, b: num}, {r: num, g: num, b: num}, ...]
+ *
+ * Use the median cut algorithm provided by quantize.js to cluster similar colors.
+ *
+ * colorCount determines the size of the palette; the number of colors returned. If not set, it
+ * defaults to 10.
+ *
+ * BUGGY: Function does not always return the requested amount of colors. It can be +/- 2.
+ *
+ * quality is an optional argument. It needs to be an integer. 1 is the highest quality settings.
+ * 10 is the default. There is a trade-off between quality and speed. The bigger the number, the
+ * faster the palette generation but the greater the likelihood that colors will be missed.
+ *
+ *
+ */
+ColorThief.prototype.getPalette = function (
+	sourceImage,
+	colorCount,
+	quality
+) {
+	const options = validateOptions( {
+		colorCount,
+		quality,
+	} );
+
+	// Create custom CanvasImage object
+	const image = new CanvasImage( sourceImage );
+	const imageData = image.getImageData();
+	const pixelCount = image.width * image.height;
+
+	const pixelArray = createPixelArray(
+		imageData.data,
+		pixelCount,
+		options.quality
+	);
+
+	// Send array to quantize function which clusters values
+	// using median cut algorithm
+	const cmap = quantize( pixelArray, options.colorCount );
+	const palette = cmap ? cmap.palette() : null;
+
+	return palette;
+};
+
+ColorThief.prototype.getImageData = function ( imageUrl, callback ) {
+	const xhr = new XMLHttpRequest();
```

Comment:
Would it be possible to avoid a network request and process the content of file when it's uploaded? In some cases like drag & drop or when uploading with the media placeholder (the one in Gutenberg) we already use data of the image to show the placeholder. /cc @ellatrix

Reply:
Tapping into `getMedia` could be good.

Reply:
I tried to not touch this file 🙈 . I'm open to changing this implementation if it's a concern